ibc-testkit/src/testapp/ibc/core/core_ctx.rs

```diff
@@ -298,46 +268,449 @@ impl ValidationContext for MockContext {
+                    None
+                }
+            })
+            .map(|consensus_path| {
+                let height = Height::new(
+                    consensus_path.revision_number,
+                    consensus_path.revision_height,
+                )?;
+                let client_state = self
+                    .consensus_state_store
+                    .get(StoreHeight::Pending, &consensus_path)
+                    .ok_or({
+                        ClientError::ConsensusStateNotFound {
+                            client_id: consensus_path.client_id,
+                            height,
+                        }
+                    })?;
+                Ok((height, client_state))
+            })
+            .collect()
+    }
+
+    /// Returns the list of heights at which the consensus state of the given client was updated.
+    fn consensus_state_heights(&self, client_id: &ClientId) -> Result<Vec<Height>, ContextError> {
+        let path = format!("clients/{}/consensusStates", client_id)
+            .try_into()
+            .map_err(|_| ClientError::Other {
+                description: "Invalid consensus state path".into(),
+            })?;
+
+        self.consensus_state_store
+            .get_keys(&path)
+            .into_iter()
+            .filter_map(|path| {
+                if let Ok(Path::ClientConsensusState(consensus_path)) = path.try_into() {
+                    Some(consensus_path)
+                } else {
+                    None
+                }
+            })
+            .map(|consensus_path| {
+                Ok(Height::new(
+                    consensus_path.revision_number,
+                    consensus_path.revision_height,
+                )?)
+            })
+            .collect::<Result<Vec<_>, _>>()
+    }
+
+    /// Connections queries all the IBC connections of a chain.
+    fn connection_ends(&self) -> Result<Vec<IdentifiedConnectionEnd>, ContextError> {
+        let path = "connections".to_owned().into();
+
+        self.connection_end_store
+            .get_keys(&path)
+            .into_iter()
+            .filter_map(|path| {
+                if let Ok(Path::Connection(connection_path)) = path.try_into() {
+                    Some(connection_path)
+                } else {
+                    None
+                }
+            })
+            .map(|connection_path| {
+                let connection_end = self
+                    .connection_end_store
+                    .get(StoreHeight::Pending, &connection_path)
+                    .ok_or_else(|| ConnectionError::ConnectionNotFound {
+                        connection_id: connection_path.0.clone(),
+                    })?;
+                Ok(IdentifiedConnectionEnd {
+                    connection_id: connection_path.0,
+                    connection_end,
+                })
+            })
+            .collect()
+    }
+
+    /// ClientConnections queries all the connection paths associated with a client.
```

Comment:
Similar with all the docstrings for these methods, they're written in such as a way as to be referencing something, `ClientConnections` in this case, that is performing the querying. But it's not obvious what is being referenced.

Reply:
b8db190d